### PR TITLE
[CALCITE] Add support for TOP N operator in SELECT statement

### PR DESCRIPTION
**Pull request by dstekol**

### CODE REVIEW

File: babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+import java.util.List;
+
+/**
+ * <code>SqlSelectTopN</code> is a class that handles the TOP n syntax in
+ * SELECT statements. It is used by the {@link SqlSelect} class.
+ */
+public class SqlSelectTopN extends SqlCall {
+  private static final SqlSpecialOperator OPERATOR =
+      new SqlSpecialOperator("TOP", SqlKind.OTHER);
```
Comment (AndrewPochapsky) [inline babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java:39]: nit: I believe @dasch-1 said that we should add new elements to SqlKind from now on.

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 2]
```diff
@@ -949,7 +949,16 @@ public void unparseOffsetFetch(SqlWriter writer, SqlNode offset,
    * @param offset Number of rows to skip before emitting, or null
    * @param fetch Number of rows to fetch, or null
    */
-  public void unparseTopN(SqlWriter writer, SqlNode offset, SqlNode fetch) {
+  public void unparseFetchAsTopN(SqlWriter writer, SqlNode offset, SqlNode fetch) {
+    //fetch.unparse(writer, -1, -1);
```
Comment (AndrewPochapsky) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java:953]: Why is this commented out? Is this case no longer necessary?

File: core/src/main/codegen/templates/Parser.jj [hunk 3]
```diff
@@ -1332,9 +1337,10 @@ SqlSelect SqlSelect() :
         }
     )
     {
-        return new SqlSelect(s.end(this), keywordList,
+        return new SqlSelect(s.end(this), keywordList, topN,
             new SqlNodeList(selectList, Span.of(selectList).pos()),
-            fromClause, where, groupBy, having, qualify, windowDecls, null, null, null,
+            fromClause, where, groupBy, having, qualify, windowDecls,
+            /*orderBy=*/null, /*offset=*/null, /*fetch=*/null,
```
Comment (AndrewPochapsky) [inline core/src/main/codegen/templates/Parser.jj:1343]: nit: Add spaces between \*/ and null

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java (hunk 2 above)
Comment (dstekol) [reply core/src/main/java/org/apache/calcite/sql/SqlDialect.java:953]: My bad. Yes, surprisingly enough the existing code relies on this method doing nothing.

File: babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java (hunk 1 above)
Comment (dstekol) [reply babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java:39]: Will do.

File: babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import org.apache.calcite.sql.SqlSelect;
+import org.apache.calcite.sql.SqlSpecialOperator;
+import org.apache.calcite.sql.SqlWriter;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.util.ImmutableNullableList;
+
+import java.util.List;
+
+/**
+ * <code>SqlSelectTopN</code> is a class that handles the TOP n syntax in
```
Comment (ArthurInTheShell) [inline babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java:34]: nit: use {@code} wrap for the class name

File: core/src/main/codegen/templates/Parser.jj [hunk 5]
```diff
@@ -1268,6 +1268,7 @@ SqlSelect SqlSelect() :
 {
     final List<SqlLiteral> keywords = new ArrayList<SqlLiteral>();
     final SqlNodeList keywordList;
+    SqlNode topN = null;
```
Comment (ArthurInTheShell) [inline core/src/main/codegen/templates/Parser.jj:1271]: nit: wrap in <#if></#if>

File: core/src/main/java/org/apache/calcite/sql/SqlSelect.java [hunk 6]
```diff
@@ -111,7 +131,8 @@ public SqlSelect(SqlParserPos pos,
 
   @Override public List<SqlNode> getOperandList() {
     return ImmutableNullableList.of(keywordList, selectList, from, where,
-        groupBy, having, qualify, windowDecls, orderBy, offset, fetch, hints);
+        groupBy, having, qualify, windowDecls, orderBy, offset, fetch,
+        hints, topN);
```
Comment (ArthurInTheShell) [inline core/src/main/java/org/apache/calcite/sql/SqlSelect.java:135]: It seems to me in the design doc, the discussion of the order of the operand does not align with the change you have here. You would either need to update the design doc, or move topN to the correct order in the code.

Comment (dstekol) [reply core/src/main/java/org/apache/calcite/sql/SqlSelect.java:135]: I will update the design doc. It seems the order is not crucial to unparsing, so I am going with Drew's suggestion of minimizing the chances of breaking anything by adding to the end.

File: babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  @Override public List<SqlNode> getOperandList() {
+    return ImmutableNullableList.of(selectNum, isPercent, withTies);
+  }
+
+  @Override public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
+    writer.keyword("TOP");
+    selectNum.unparse(writer, leftPrec, rightPrec);
+    if (isPercent.getValueAs(Boolean.class)) {
```
Comment (dasch-1) [inline babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java:63]: Should we also handle the case where isPercent is null?

File: babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+      writer.keyword("PERCENT");
+    }
+    if (withTies.getValueAs(Boolean.class)) {
```
Comment (dasch-1) [inline babel/src/main/java/org/apache/calcite/sql/babel/SqlSelectTopN.java:66]: Should we also handle the case where withTies is null?

File: core/src/main/codegen/templates/Parser.jj [hunk 9]
```diff
@@ -1307,6 +1308,10 @@ SqlSelect SqlSelect() :
     |   <ALL> {
             keywords.add(SqlSelectKeyword.ALL.symbol(getPos()));
         }
+<#if parser.includeTopN>
+    |
+        topN = SqlSelectTopN(getPos())
```
Comment (dasch-1) [inline core/src/main/codegen/templates/Parser.jj:1313]: So does this mean that the TOP N clause is mutually exclusive with the DISTINCT keyword?

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java [hunk 10]
```diff
@@ -949,7 +949,15 @@ public void unparseOffsetFetch(SqlWriter writer, SqlNode offset,
...
    */
-  public void unparseTopN(SqlWriter writer, SqlNode offset, SqlNode fetch) {
+  public void unparseFetchAsTopN(SqlWriter writer, SqlNode offset, SqlNode fetch) {
+  }
+
+  /**
+   * Unparses a TOP N clause (ex. SELECT TOP 5 * FROM FOO)
+   * @param topN
+   */
+  public void unparseTopN(SqlWriter writer, SqlNode topN) {
```
Comment (dasch-1) [inline core/src/main/java/org/apache/calcite/sql/SqlDialect.java:959]: Should we check for a null topN here?

File: core/src/main/codegen/templates/Parser.jj (hunk 9 above)
Comment (dstekol) [reply core/src/main/codegen/templates/Parser.jj:1313]: Yes, this is what the syntax diagram says the behavior should be.

File: core/src/main/java/org/apache/calcite/sql/SqlDialect.java (hunk 10 above)
Comment (dstekol) [reply core/src/main/java/org/apache/calcite/sql/SqlDialect.java:959]: This check is done in SqlSelectOperator, but the unparseTopN could theoretically be called from somewhere else as well, so it probably does make sense to add it just to be safe.

File: babel/src/main/codegen/includes/parserImpls.ftl [hunk 11]
```diff
@@ -1197,3 +1197,42 @@ SqlNode RankFunctionCallWithParams() :
         return SqlStdOperatorTable.OVER.createCall(s.end(over), rankCall, over);
     }
 }
+
```
Comment (lisayin) [inline babel/src/main/codegen/includes/parserImpls.ftl:1200]: Maybe add a top level comment?

File: babel/src/test/java/org/apache/calcite/test/BabelParserTest.java [hunk 12]
```diff
@@ -1667,4 +1667,45 @@ private void checkParseInfixCast(String sqlType) {
...
+  }
+
+  @Test void testTopNDecimalWithoutPercentFails() {
+    final String sql = "select top ^5.2^ bar from foo";
+    final String expected = "(?s).*Cannot specify non-integer value."
+        + "*without specifying PERCENT.*";
+    sql(sql).fails(expected);
+  }
+
+  @Test void testTopNPercentGreaterThan100Fails() {
```
Comment (lisayin) [inline babel/src/test/java/org/apache/calcite/test/BabelParserTest.java:1706]: Add a test that sets N to be more than 100 but doesn't specify PERCENT

File: core/src/main/java/org/apache/calcite/sql/dialect/SybaseSqlDialect.java [hunk 13]
```diff
@@ -36,11 +36,11 @@ public SybaseSqlDialect(Context context) {
 
   @Override public void unparseOffsetFetch(SqlWriter writer, SqlNode offset,
       SqlNode fetch) {
-    // No-op; see unparseTopN.
+    // No-op; see unparseFetchAsTopN.
```
Comment (lisayin) [inline core/src/main/java/org/apache/calcite/sql/dialect/SybaseSqlDialect.java:39]: I might have missed it but where is FETCH NEXT n ROWS -> SELECT TOP(n) tested?

Comment (dstekol) [reply core/src/main/java/org/apache/calcite/sql/dialect/SybaseSqlDialect.java:39]: I believe this is tested in the RelToSqlConverterTest file (for example in testSelectQueryWithFetchClause unit test). Converting FETCH statements into SELECT TOP N statements is not actually the functionality we are trying to add here, so the changes here are just to disentangle the TOP N unparsing logic from the FETCH unparsing logic (which sometimes ends up being unparsed as a TOP N statement as per pre-existing code).

Comment (lisayin) [reply core/src/main/java/org/apache/calcite/sql/dialect/SybaseSqlDialect.java:39]: That makes sense

File: babel/src/test/java/org/apache/calcite/test/BabelParserTest.java [hunk 14]
```diff
@@ -1667,4 +1667,52 @@ private void checkParseInfixCast(String sqlType) {
...
+    final String sql = "select top 200 ^percent^ bar from foo";
+    final String expected = "(?s).*Numeric literal.*out of range.*";
+    sql(sql).fails(expected);
+  }
```
Comment (michaelvdow) [inline babel/src/test/java/org/apache/calcite/test/BabelParserTest.java:1717]: Could add a test here for negative numbers to show that it should fail on that input

File: core/src/main/java/org/apache/calcite/sql/SqlKind.java [hunk 15]
```diff
@@ -1007,6 +1007,9 @@
   /** The {@code TUMBLE} group function. */
   TUMBLE,
 
+  /** The {@code TOP N} function*/
```
Comment (michaelvdow) [inline core/src/main/java/org/apache/calcite/sql/SqlKind.java:1010]: nit: period and space after "function"

File: core/src/main/java/org/apache/calcite/sql/SqlWriter.java [hunk 16]
```diff
@@ -376,11 +376,16 @@ public String getName() {
...
-   * Prints the TOP(n) clause.
+   * Prints the fetch clause as a TOP(n) clause.
    *
    * @see #fetchOffset
    */
-  void topN(SqlNode fetch, SqlNode offset);
+  void fetchAsTopN(SqlNode fetch, SqlNode offset);
+
+  /**
+   * Prints the TOP N clause
```
Comment (michaelvdow) [inline core/src/main/java/org/apache/calcite/sql/SqlWriter.java:386]: nit: period

File: babel/src/test/java/org/apache/calcite/test/BabelParserTest.java (hunk 14 above)
Comment (dstekol) [reply babel/src/test/java/org/apache/calcite/test/BabelParserTest.java:1717]: Good call, will do.
